# The First Data Analysis Project using pandas

The dataset was originally scraped and uploaded to Kaggle. It has been made modified from the original dataset that was uploaded to Kaggle:
- 50,000 data points has been sampled from the full dataset, to ensure the code runs quickly in a hosted environment
- The dataset has been dirtied a bit in order to closely resemble a real world example (the version uploaded to Kaggle was cleaned to be easier to work with)

**The aim of this project is to clean the data and analyze the included used car listings. In addition, to become familiar with some of the unique benefits jupyter notebook provides for pandas.**

First, let's import the pandas and NumPy libraries

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [6]:
autos.head(10)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

As clearly seen in the cell above, we can see following observations:
- The dataset contains 20 columns, most of them are strings
- Some columns have null values, overall percentage of null values are less than ~20%
- The columns names are defined using camalCase, not snake_case
- Most of the values are written in German

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [3]:
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen']
autos.columns = new_columns
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

The data dictionary provided with data is as follows:

| index_number | Label               | Description                                                                |
|--------------|---------------------|----------------------------------------------------------------------------|
| 0            | date_crawled        | When this ad was first crawled. All field-values are taken from this date. |
| 1            | name                | Name of the car.                                                           |
| 2            | seller              | Whether the seller is private or a dealer.                                 |
| 3            | offer_type          | The type of listing                                                        |
| 4            | price               | The price on the ad to sell the car.                                       |
| 5            | ab_test             | Whether the listing is included in an A/B test.                            |
| 6            | vehicle_type        | The type of the vehicle                                                    |
| 7            | registration_year   | The year in which the car was first registered.                            |
| 8            | gearbox             | The transmission type.                                                     |
| 9            | power_ps            | The power of the car in PS.                                                |
| 10           | model               | The car model name.                                                        |
| 11           | odometer            | How many kilometers the car has driven.                                    |
| 12           | registration_month  | The month in which the car was first registered.                           |
| 13           | fuel_type           | What type of fuel the car uses.                                            |
| 14           | brand               | The brand of the car.                                                      |
| 15           | unrepaired_damage   | If the car has a damage which is not yet repaired.                         |
| 16           | ad_created          | The date on which the eBay listing was created.                            |
| 17           | number_of_pictures  | The number of pictures in the ad.                                          |
| 18           | postal_code         | The postal code for the location of the vehicle.                           |
| 19           | last_seen           | When the crawler saw this ad last online.                                  |

Secondly, we need to convert German words into English words
1. Create a map with german words and their respective words in English
2. Apply to our dataset

In [8]:
print(autos["seller"].unique())
print(autos["ab_test"].unique())
print(autos["vehicle_type"].unique())
print(autos["gearbox"].unique())
print(autos["fuel_type"].unique())
print(autos["unrepaired_damage"].unique())

['privat' 'gewerblich']
['control' 'test']
['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']
['manuell' 'automatik' nan]
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']
['nein' nan 'ja']


In [4]:
english_words = {"privat":"private",
                 "gewerblich":"commercial",
                 "bus":"bus",
                 "kombi":"microbus",
                 "limousine":"limousine",
                 "kleinwagen":"supermini",
                 "coupe":"coupe",
                 "suv":"suv",
                 "cabrio":"convertible",
                 "andere":"other",
                 "manuell":"manual",
                 "automatik":"automatic",
                 "lpg":"lpg",
                 "benzin":"gas",
                 "diesel":"diesel",
                 "cng":"cng",
                 "hybrid":"hybrid",
                 "elektro":"electro",
                 "nein":"no",
                 "ja":"yes"}
autos["seller"] = autos["seller"].map(english_words)
autos["ab_test"] = autos["ab_test"].map(english_words)
autos["vehicle_type"] = autos["vehicle_type"].map(english_words)
autos["gearbox"] = autos["gearbox"].map(english_words)
autos["fuel_type"] = autos["fuel_type"].map(english_words)
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(english_words) # use it only once

Thirdly, look for 3 types of columns:
- Any columns that have mostly one value that are candidates to be dropped
- Any columns that need more investigation.

| index_number | Label               | Description                                                                |
|--------------|---------------------|----------------------------------------------------------------------------|
| 0            | date_crawled        | When this ad was first crawled. All field-values are taken from this date. |
| 16           | ad_created          | The date on which the eBay listing was created.                            |
| 19           | last_seen           | When the crawler saw this ad last online.                                  |

- Any examples of numeric data stored as text that needs to be cleaned.

| index_number | Label               | Description                                                                |
|--------------|---------------------|----------------------------------------------------------------------------|
| 4            | price               | The price on the ad to sell the car.                                       |
| 11           | odometer            | How many kilometers the car has driven.                                    |

1. Let's remove any non-numeric characters
2. Convert the columns to a numeric values
3. Rename the columns accordingly

In [5]:
autos["odometer"] = autos["odometer"].str.strip().str.replace("km", "").str.replace(",", "").astype(int)
autos["price"] = autos["price"].str.strip().str.replace("$", "").str.replace(",", "").astype(int)
autos.rename(columns={"price":"price_dollars", "odometer":"odometer_km"}, inplace=True)

We will drop a few columns (ab_test, seller and offer type) because they do not provide any usefull information since the results are monothonic.

In [6]:
autos = autos.drop(["ab_test", "seller", "offer_type"], axis=1)

In [12]:
autos.head(10)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   price_dollars vehicle_type  registration_year    gearbox  power_ps  \
0           5000          bus               2004     manual       158   
1           8500    limousine               1997  automatic       286   
2           8990    limousine               2009     manual       102   
3           4350    supermini               2007  automatic        71   
4           1350     microbus               2003     manual         0   
5           7900          bus               2006  automatic       150   
6            300    limousine               1995     manual        90   
7           1990    limousine               1998     manual        90   
8            250          NaN               2000     manual         0   
9            590          bus               1997     manual        90   

     model  odometer_km  registration_month fuel_type       brand  \
0   andere       150000                   3       lpg     peugeot   
1      7er       150000                   6       gas         bmw   
2     golf        70000                   7       gas  volkswagen   
3   fortwo        70000                   6       gas       smart   
4    focus       150000                   7       gas        ford   
5  voyager       150000                   4    diesel    chrysler   
6     golf       150000                   8       gas  volkswagen   
7     golf       150000                  12    diesel  volkswagen   
8    arosa       150000                  10       NaN        seat   
9   megane       150000                   7       gas     renault   

  unrepaired_damage           ad_created  number_of_pictures  postal_code  \
0                no  2016-03-26 00:00:00                   0        79588   
1                no  2016-04-04 00:00:00                   0        71034   
2                no  2016-03-26 00:00:00                   0        35394   
3                no  2016-03-12 00:00:00                   0        33729   
4                no  2016-04-01 00:00:00                   0        39218   
5               NaN  2016-03-21 00:00:00                   0        22962   
6               NaN  2016-03-20 00:00:00                   0        31535   
7                no  2016-03-16 00:00:00                   0        53474   
8                no  2016-03-22 00:00:00                   0         7426   
9                no  2016-03-16 00:00:00                   0        15749   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50  
5  2016-04-06 09:45:21  
6  2016-03-23 02:48:59  
7  2016-04-07 03:17:32  
8  2016-03-26 18:18:10  
9  2016-04-06 10:46:35

Let's continue exploring the data, specifically looking for data that doesn't look right. Start by analyzing the odometer_km and price columns. Here's the steps:

1. Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

Use:
- Series.unique().shape to see how many unique values
- Series.describe() to view min/max/median/mean etc
- Series.value_counts(), with some variations: *chained to .head() if there are lots of values.*

Because Series.value_counts() returns a series, we can use Series.sort_index() with ascending= True or False to view the highest and lowest values with their counts (can also chain to head() here).

When removing outliers, we can do **df[(df["col"] > x ) & (df["col"] < y )]**, but it's more readable to use **df[df["col"].between(x,y)]**

In [13]:
autos["price_dollars"].unique().shape

(2357,)

In [14]:
autos["price_dollars"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64

In [15]:
autos["price_dollars"].value_counts().sort_index(ascending=True).head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price_dollars, dtype: int64

In the cells above, we can see the "price_dollars" column has 2357 unique values of price, with the some really strange outliers such as 0 (free cars) and (9999999+ dollars for car), which is absolute nonsense. By removing outliers, we can make our dataset more clean in order to provide accurate data analysis.

To be persistent, we will remove rows with the price lower than \\$1000 and higher than \\$100000.

In [7]:
average_autos = autos["price_dollars"].between(1000, 100000, inclusive=True)
autos = autos[average_autos]
autos.shape

(38587, 17)

In [17]:
autos["odometer_km"].unique().shape

(13,)

In [18]:
autos["odometer_km"].describe()

count     38587.000000
mean     122859.771426
std       40712.638070
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [8]:
autos["odometer_km"].value_counts().sort_index(ascending=True).head(13)

5000        498
10000       225
20000       684
30000       742
40000       794
50000       984
60000      1097
70000      1154
80000      1333
90000      1569
100000     1858
125000     4338
150000    23311
Name: odometer_km, dtype: int64

**The values of the "odometer_km" column look acceptable. There are not strange outliers**

Let's now move on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

3 of these dates are identified as strings values by pandas. Because there are 3 columns representing dates in a string format, we need to convert it into a numerical format so we can understand it quantitatively. The others 2 values already represented as numeric values. 

In [20]:
autos[["date_crawled", "last_seen", "ad_created"]][0:5]

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

The date format of all 3 columns is yyyy-mm-dd HH:MM:SS, however the "ad_created" columns has no time inputs. So the date would be formated according to yyyy-mm-dd format, without the specific time.

In [9]:
autos["date_crawled"].value_counts(normalize=True, dropna=False).sort_index(ascending=True).head(5)

2016-03-05 14:06:30    0.000026
2016-03-05 14:06:40    0.000026
2016-03-05 14:07:21    0.000026
2016-03-05 14:07:26    0.000026
2016-03-05 14:07:40    0.000026
Name: date_crawled, dtype: float64

In the cell above, we can see that the distribution of values in percentages, the mode is 0.000026 and the average is 0.000032.

In [22]:
autos["registration_year"].describe()

count    38587.000000
mean      2005.680125
std         86.727780
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

There are 38587 different values of registration_year, the earliest registered car dated in 1000 (which is outlier) and the latest registered in 9999 (which in not possible), all values less than 1960 and 2016 will be removed for better readability.

In [10]:
autos_registered = autos["registration_year"].between(1960, 2016, inclusive=True)
autos = autos[autos_registered]
autos.shape

(37129, 17)

In [11]:
autos["registration_year"].value_counts(normalize=True).sort_index(ascending=True)

1960    0.000458
1961    0.000162
1962    0.000108
1963    0.000215
1964    0.000242
1965    0.000458
1966    0.000566
1967    0.000673
1968    0.000673
1969    0.000485
1970    0.000862
1971    0.000646
1972    0.000808
1973    0.000619
1974    0.000619
1975    0.000485
1976    0.000566
1977    0.000512
1978    0.001104
1979    0.000835
1980    0.001912
1981    0.000700
1982    0.001023
1983    0.001347
1984    0.001266
1985    0.002128
1986    0.001508
1987    0.001374
1988    0.002828
1989    0.003313
1990    0.005090
1991    0.005575
1992    0.005844
1993    0.005871
1994    0.007299
1995    0.011904
1996    0.014598
1997    0.021169
1998    0.034474
1999    0.046433
2000    0.053866
2001    0.055590
2002    0.057502
2003    0.066713
2004    0.070242
2005    0.075009
2006    0.071400
2007    0.060815
2008    0.059307
2009    0.055913
2010    0.042554
2011    0.043524
2012    0.035094
2013    0.021304
2014    0.017426
2015    0.009615
2016    0.017372
Name: registration_year, dtype:

After the dataset has been filtered according to the price, we had 38587 rows. Now, after filtering the registration year, we have 37129 rows. The values mostly distributed between 1998 and 2012 years.

In [25]:
autos["brand"].value_counts().head(10)

volkswagen       7841
bmw              4660
mercedes_benz    4142
audi             3630
opel             3312
ford             2177
renault          1387
peugeot          1038
fiat              784
skoda             709
Name: brand, dtype: int64

I have chosen the most frequent brands (top-10 listed above) and will look for their mean price.

In [12]:
brands = autos["brand"].value_counts().head(10).index
print(brands)

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford', 'renault',
       'peugeot', 'fiat', 'skoda'],
      dtype='object')


In [62]:
popular_autos = []

for index, row in autos.iterrows():
    brand = row[11]
    price = row[2]
    odometer = row[8]
    if brand in brands:
        popular_autos.append([brand, price, odometer])

autos_sum_price = {}
autos_sum_odometer = {}
autos_count = {}

for value in popular_autos:
    brand = value[0]
    price = int(value[1])
    odometer = int(value[2])
    if brand in autos_count:
        autos_count[brand] += 1
        autos_sum_price[brand] += price
        autos_sum_odometer[brand] += odometer
    else:
        autos_count[brand] = 1
        autos_sum_price[brand] = price
        autos_sum_odometer[brand] = odometer

autos_mean_price = {}
autos_mean_odometer = {}

for brand in autos_count:
    avg_price = autos_sum_price[brand]/autos_count[brand]
    avg_odometer = autos_sum_odometer[brand]/autos_count[brand]
    autos_mean_price[brand] = avg_price
    autos_mean_odometer[brand] = avg_odometer

In the cell above, the mean values of "price" and "odometer_km" are evaluated and defined in dictionaries. In the next 2 cells, the dictionaries are transformed into pandas.Series and merged together as a dataframe.

In [67]:
bmp = pd.Series(autos_mean_price)
bmo = pd.Series(autos_mean_odometer)
bmo

peugeot          122341.040462
bmw              132095.493562
volkswagen       125765.846193
ford             119839.228296
audi             127524.793388
renault          121423.936554
mercedes_benz    130245.050700
opel             123996.074879
skoda            110063.469676
fiat             107901.785714
dtype: float64

In [68]:
bmv = {"mean_price": bmp, "mean_odometer": bmo}
brand_values = pd.DataFrame(bmv)
brand_values

mean_price  mean_odometer
peugeot         3955.169557  122341.040462
bmw             8974.562017  132095.493562
volkswagen      6640.999235  125765.846193
ford            5224.046394  119839.228296
audi           10276.903581  127524.793388
renault         3590.942322  121423.936554
mercedes_benz   9155.143892  130245.050700
opel            4218.230676  123996.074879
skoda           6836.696756  110063.469676
fiat            4008.174745  107901.785714

It clearly visible from the above that audi, mercedes_benz and bmw are the most expensive and the most durable cars since their mean_price and mean_odometer values are the highest as compared to the others. Opel, renault and fiat are the cheapest cars and their long-standing is questionable. Volkswagen and skoda somewhere between, however volkswagen cars are exploited longer compare to skoda cars.